network/src/connectivity_manager/mod.rs

```diff
@@ -394,22 +407,26 @@ where
         }
     }
 
-    async fn dial_eligible_peers<'a>(
+    fn dial_eligible_peers<'a>(
         &'a mut self,
         pending_dials: &'a mut FuturesUnordered<BoxFuture<'static, PeerId>>,
     ) {
-        let eligible = self.eligible.read().clone();
+        let to_connect = self.choose_peers_to_dial();
+        for (peer_id, peer) in to_connect {
+            self.queue_dial_peer(peer_id, peer, pending_dials);
+        }
+    }
+
+    fn choose_peers_to_dial(&mut self) -> Vec<(PeerId, DiscoveredPeer)> {
```

Comment:
if you're feeling extra ambitious, you could return `&'a DiscoveredPeer` here : p

Reply:
I'm not feeling like dealing with passing around all the lifetimes right now, keeping it simple for now.